src/cargo/core/resolver/context.rs

```diff
@@ -246,10 +248,13 @@ impl Context {
 
             let set = self.resolve_features
                 .entry(pkgid.clone())
-                .or_insert_with(HashSet::new);
+                .or_insert_with(|| Rc::new(HashSet::new()));
+
+            let mut inner: HashSet<_> = (**set).clone();
```

Comment:
Could this perhaps use `Rc::make_mut`? That way we could have a fast path as well for when the `Context` hasn't been cloned?

Reply:
This is an algorithmic improvement, I think `HashSet.clone()` is now `O(# activated * ticks)` and this pr bings it to `O(# activated)`.

I can experiment with that, here and for #5118, but I suspect it will be rare. Not only does the context need not to be cloned, but it needs not to have been cloned since last time that this package was added. Would you like me to do that expremint in this PR or in a follow up?

Reply:
I did it in this PR as it made the diff much smaller. New performance numbers on the way.